File: SICP/Lecture_1/Code/Erlang/fixedpoint.erl
-module(fixedpoint).
-compile(export_all).

-define(TOLERANCE, 0.001).

fixedpoint(F) ->
    fun(Guess) ->
        fun(X) ->
            try_once(Guess, F, X)
        end
    end.

try_once(Guess, F, X) ->
    NextGuess = (F(Guess))(X),
    case good_enough(NextGuess, Guess) of
        true ->
            NextGuess;
        false ->
            try_once(NextGuess, F, X)
    end.

good_enough(NextGuess, Guess) ->
    abs(NextGuess - Guess) < ?TOLERANCE.

test_sqrt() ->
    F = fun(Y) ->
            fun(X) ->
                (Y + X / Y) / 2
            end
        end,
    Sqrt = (fixedpoint(F))(1),
    true = abs(Sqrt(4) - 2) < ?TOLERANCE,
    true = abs(Sqrt(9) - 3) < ?TOLERANCE,
    test_sqrt_ok.

test_cuberoot() ->
    F = fun(Y) ->
            fun(X) ->
                (X / (Y * Y) + 2 * Y) / 3
            end
        end,
    Cuberoot = (fixedpoint(F))(1),
    true = abs(Cuberoot(8) - 2) < ?TOLERANCE,
    true = abs(Cuberoot(27) - 3) < ?TOLERANCE,
    test_cuberoot_ok.

test() ->
    test_sqrt(),
    test_cuberoot(),
    test_ok.

File: SICP/Lecture_1/Code/Python/sqrt/fixed_point.py
TOLERANCE = 0.001

def sqrt(x):
    func = lambda y: lambda x: (y + x / y) / 2.0
    sqrt_func = (fixed_point(func))(1.0)
    return sqrt_func(x)

def cube_root(x):
    func = lambda y: lambda x: (x / (y * y) + 2 * y) / 3.0
    cube_root_func = (fixed_point(func))(1.0)
    return cube_root_func(x)

def fixed_point(func):
    return lambda guess: lambda x: try_once(guess, func, x)

def try_once(guess, func, x):
    next_guess = (func(guess))(x);
    if good_enough(next_guess, guess):
        return next_guess
    else:
        return try_once(next_guess, func, x)

def good_enough(next_guess, guess):
    return abs(next_guess - guess) < TOLERANCE

def average(x, y):
    return (x + y) / 2.0